velox/dwio/parquet/reader/NestedStructureDecoder.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/dwio/parquet/reader/NestedStructureDecoder.h"
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/dwio/common/BufferUtil.h"
+
+namespace facebook::velox::parquet {
+
+int64_t NestedStructureDecoder::readOffsetsAndNulls(
+    const uint8_t* definitionLevels,
+    const uint8_t* repetitionLevels,
+    int64_t numValues,
+    uint8_t maxDefinition,
+    uint8_t maxRepeat,
+    BufferPtr& offsetsBuffer,
+    BufferPtr& lengthsBuffer,
+    BufferPtr& nullsBuffer,
+    memory::MemoryPool& pool) {
+  dwio::common::ensureCapacity<uint64_t>(
+      nullsBuffer, numValues / 64 + 1, &pool);
+  dwio::common::ensureCapacity<vector_size_t>(
+      offsetsBuffer, numValues + 1, &pool);
+  dwio::common::ensureCapacity<vector_size_t>(lengthsBuffer, numValues, &pool);
+
+  auto offsets = offsetsBuffer->asMutable<vector_size_t>();
+  auto lengths = lengthsBuffer->asMutable<vector_size_t>();
+  auto nulls = nullsBuffer->asMutable<uint64_t>();
+
+  int64_t lastCollectionEnd = 0;
+  bool wasLastCollectionNull = definitionLevels[0] == (maxDefinition - 1);
+  bits::setBit<uint64_t>(nulls, 0, wasLastCollectionNull);
+  offsets[0] = 0;
+
+  int64_t outputIndex = 1;
+  for (int64_t i = 1; i < numValues; ++i) {
+    uint8_t definitionLevel = definitionLevels[i];
+    uint8_t repetitionLevel = repetitionLevels[i];
+
+    bool isNull = definitionLevel == (maxDefinition - 1);
+    bits::setBit<uint64_t>(nulls, outputIndex, isNull);
+
+    auto length = (i - lastCollectionEnd) * !wasLastCollectionNull;
+    lengths[outputIndex - 1] = length;
+    offsets[outputIndex] = offsets[outputIndex - 1] + length;
+
+    // Always update the outputs, but only increase the outputIndex when the
+    // current entry is the begin of a new collection, and it's not empty (empty
+    // means it belongs to a row that is null in one of the ancestor levels)
+    bool isNotEmpty = definitionLevel >= (maxDefinition - 1);
+    bool isCollectionBegin = repetitionLevel < maxRepeat;
+    outputIndex += isCollectionBegin && isNotEmpty;
+    wasLastCollectionNull = isNotEmpty ? isNull : wasLastCollectionNull;
+    lastCollectionEnd = isCollectionBegin ? i : lastCollectionEnd;
+  }
+
+  auto length = (numValues - lastCollectionEnd) * !wasLastCollectionNull;
+  lengths[outputIndex - 1] = length;
+  offsets[outputIndex] = offsets[outputIndex - 1] + length;
```

Comment:
Do we need this?

Reply:
@Yuhta I'm not sure what you meant? Are you asking if we need to output the offsets at all? I see `ArrayVector` needs both lengths and offsets so I calculate offsets too.

Reply:
I mean the last `offsets` is probably not needed, we only need `numValues` of them